File: javascript/src/euler/solution089.js
// https://projecteuler.net/problem=89
// Roman numerals
//
// The rules for writing Roman numerals allow for many ways of writing each number (see About Roman Numerals...). However, there is always a "best" way of writing a particular number.
//
// For example, the following represent all of the legitimate ways of writing the number sixteen:
//
// IIIIIIIIIIIIIIII
// VIIIIIIIIIII
// VVIIIIIII
// XIIIIII
// VVVI
// XVI
//
// The last example being considered the most efficient, as it uses the least number of numerals.
//
// The 11K text file, roman.txt (right click and 'Save Link/Target As...'), contains one thousand numbers written in valid, but not necessarily minimal, Roman numerals; that is, they are arranged in descending units and obey the subtractive notation (see About Roman Numerals... for the definitive rules for this problem).
//
// Find the number of characters saved by writing each of these in their minimal form.
//
// Note: You can assume that all the Roman numerals in the file contain no more than four consecutive identical units.
//
// The final answer to the problem is 743.
//
// Execution time measurements: Approximately 10 ms.
//
// Details on any optimizations made during development: Direct parsing and conversion using value tables, no complex logic.
// Interesting notes or insights from the AI implementation process: Adapted parsing and generation from C++ using arrays for Roman values.

const fs = require('fs');

function roman_to_int(s) {
  const values = [
    ['M', 1000], ['CM', 900], ['D', 500], ['CD', 400],
    ['C', 100], ['XC', 90], ['L', 50], ['XL', 40],
    ['X', 10], ['IX', 9], ['V', 5], ['IV', 4], ['I', 1]
  ];
  let total = 0;
  let i = 0;
  while (i < s.length) {
    let found = false;
    for (const [str, val] of values) {
      if (s.substr(i, str.length) === str) {
        total += val;
        i += str.length;
        found = true;
        break;
      }
    }
    if (!found) i++;
  }
  return total;
}

function int_to_roman(num) {
  const values = [
    [1000, 'M'], [900, 'CM'], [500, 'D'], [400, 'CD'],
    [100, 'C'], [90, 'XC'], [50, 'L'], [40, 'XL'],
    [10, 'X'], [9, 'IX'], [5, 'V'], [4, 'IV'], [1, 'I']
  ];
  let result = '';
  for (const [val, str] of values) {
    while (num >= val) {
      result += str;
      num -= val;
    }
  }
  return result;
}

function roman_numerals() {
  const data = fs.readFileSync('src/euler/p089_roman.txt', 'utf8');
  const lines = data.trim().split('\n');
  let saved = 0;
  for (const line of lines) {
    if (!line) continue;
    const value = roman_to_int(line);
    const minimal = int_to_roman(value);
    saved += line.length - minimal.length;
  }
  return saved;
}

module.exports = {
  answer: () => roman_numerals()
};

File: java/src/main/java/org/tvarley/euler/solutions/Solution089.java
// https://projecteuler.net/problem=89
// Roman Numerals
//
// For a number written in Roman numerals to be considered valid there are basic rules which must be followed. Even though the rules allow some numbers to be expressed in more than one way there is always a "best" way of writing a particular number.
//
// For example, it would appear that there are at least six ways of writing the number sixteen:
//
// IIIIIIIIIIIIIIII
// VIIIIIIIIIII
// VVIIIIIII
// XIIIIII
// VVVI
// XVI
//
// However, according to the rules only XIIIIII and XVI are valid, and XIIIIII is not the "best" way of writing it, because it uses each symbol too many times.
//
// The "best" (shortest) way to write them, using the subtractive notation, is XVI.
//
// Find the number of characters saved by writing each of these in their minimal form.
//
// Note: You can assume that all the Roman numerals in the file contain no more than four consecutive identical units.
//
// Answer: 743
//
// Completed on: 2024-12-04
// Time: ~1ms
// Optimizations: Greedy parsing and building
// Notes: Used BufferedReader for file reading.

package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.List;

public class Solution089 implements Solution {
    private static final List<String[]> ROMAN_VALUES = Arrays.asList(
        new String[]{"M", "1000"}, new String[]{"CM", "900"}, new String[]{"D", "500"}, new String[]{"CD", "400"},
        new String[]{"C", "100"}, new String[]{"XC", "90"}, new String[]{"L", "50"}, new String[]{"XL", "40"},
        new String[]{"X", "10"}, new String[]{"IX", "9"}, new String[]{"V", "5"}, new String[]{"IV", "4"}, new String[]{"I", "1"}
    );

    public String solve() {
        long saved = 0;
        try (BufferedReader br = new BufferedReader(new InputStreamReader(
                getClass().getResourceAsStream("/p089_roman.txt")))) {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.isEmpty()) continue;
                int value = romanToInt(line);
                String minimal = intToRoman(value);
                saved += line.length() - minimal.length();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return Long.toString(saved);
    }

    private int romanToInt(String s) {
        int total = 0;
        int i = 0;
        while (i < s.length()) {
            boolean found = false;
            for (String[] p : ROMAN_VALUES) {
                String sym = p[0];
                int val = Integer.parseInt(p[1]);
                if (i + sym.length() <= s.length() && s.substring(i, i + sym.length()).equals(sym)) {
                    total += val;
                    i += sym.length();
                    found = true;
                    break;
                }
            }
            if (!found) i++;
        }
        return total;
    }

    private String intToRoman(int num) {
        StringBuilder result = new StringBuilder();
        for (String[] p : ROMAN_VALUES) {
            int val = Integer.parseInt(p[1]);
            String sym = p[0];
            while (num >= val) {
                result.append(sym);
                num -= val;
            }
        }
        return result.toString();
    }
}